File: Basic Programming/Hackerrank 30 days of code python3 solution/Day_22.py
#                      Name : Arif Khan
#                      Judge: HACKERRANK
#                      University: Primeasia University
#                      problem: Day 22: Binary Search Trees
#                      Difficulty: Easy
#                      Problem Link: https://www.hackerrank.com/challenges/30-binary-search-trees/problem
#

class Node:
    def __init__(self,data):
        self.right=self.left=None
        self.data = data
class Solution:
    def insert(self,root,data):
        if root==None:
            return Node(data)
        else:
            if data<=root.data:
                cur=self.insert(root.left,data)
                root.left=cur
            else:
                cur=self.insert(root.right,data)
                root.right=cur
        return root

    def getHeight(self,root):
        if root == None:
            return -1
        l = self.getHeight(root.left)
        r = self.getHeight(root.right)
        return max(l, r)+1

T=int(input())
myTree=Solution()
root=None
for i in range(T):
    data=int(input())
    root=myTree.insert(root,data)
height=myTree.getHeight(root)
print(height)       

File: Basic Programming/Hackerrank 30 days of code c++ solution/Day_22.cpp
/*                    Name : Arif Khan
                      Judge: HACKERRANK
                      University: Primeasia University
                      problem: Day 22: Binary Search Trees
                      Difficulty: Easy
                      Problem Link: https://www.hackerrank.com/challenges/30-binary-search-trees/problem
*/

#include <iostream>
#include <cstddef>

using namespace std;	

class Node{
    public:
        int data;
        Node *left;
        Node *right;
        Node(int d){
            data = d;
            left = NULL;
            right = NULL;
        }
};
class Solution{
    public:
  		Node* insert(Node* root, int data) {
            if(root == NULL) {
                return new Node(data);
            }
            else {
                Node* cur;
                if(data <= root->data){
                    cur = insert(root->left, data);
                    root->left = cur;
                }
                else{
                    cur = insert(root->right, data);
                    root->right = cur;
               }

               return root;
           }
        }

	int getHeight(Node* root){ 
            if(root == NULL)
                return -1;
        int left=getHeight(root->left);
        int right=getHeight(root->right);
        return max(left, right) + 1;
    }

}; //End of Solution

int main() {
    Solution myTree;
    Node* root = NULL;
    int t;
    int data;

    cin >> t;

    while(t-- > 0){
        cin >> data;
        root = myTree.insert(root, data);
    }
    int height = myTree.getHeight(root);
    cout << height;

    return 0;
}